src/openms/include/OpenMS/METADATA/ID/IdentificationData.h

```diff
@@ -662,8 +858,27 @@ namespace OpenMS
       }
     }
 
+    /// Helper function to add a meta value to an element in a multi-index container
+    template <typename RefType, typename ContainerType>
+    void setMetaValue_(const RefType ref, const String& key, const DataValue& value,
+                       ContainerType& container, const AddressLookup& lookup = AddressLookup())
+    {
+      if (!no_checks_ && ((lookup.empty() && !isValidReference_(ref, container)) ||
+                          (!lookup.empty() && !isValidHashedReference_(ref, lookup))))
+      {
+        String msg = "invalid reference for the given container";
+        throw Exception::IllegalArgument(__FILE__, __LINE__,
+                                         OPENMS_PRETTY_FUNCTION, msg);
+      }
+      container.modify(ref, [&key, &value](typename ContainerType::value_type& element)
+      {
+        element.setMetaValue(key, value);
+      });
+    }
+
 
-    // IDFilter needs access to do its job:
+    // these classes need access to manipulate data:
     friend class IDFilter;
+    friend class MapAlignmentTransformer;
```

Comment:
**this is IMHO a design mistake that needs to be fixed.**
IDDataStructure needs a clean/safe and efficient interface so that IDFilter and MapAlignmentTransformer can implement the modifications.
Otherwise, we can't put IDDataStructure in a library on which others build to e.g. implement custom IDFilter etc.

Reply:
This is a tricky one. IdentificationData purposely doesn't expose an interface for low-level manipulation (e.g. removing individual items of data), because consistency of the contents needs to be ensured and re-establishing that is costly (see `IdentificationData::cleanup`). Instead more high-level functions are provided (for filtering and transforming retention times) that however need privileged access to IdentificationData internals. It would be possible to move these functions into IdentificationData itself (to avoid the friend classes), but from a "separation of concerns" viewpoint I think that makes less sense.

> Otherwise, we can't put IDDataStructure in a library on which others build to e.g. implement custom IDFilter etc.

The point is that "user code" shouldn't implement filtering. It should just use the library functions that are implemented in a safe manner in IDFilter.

Reply:
IMHO That argument doesn’t fly for such a low level functionality. So how would this look like if we add functionality tools? E.g. PeptideIndexer? Or some of the inference algorithms? How can there be implemented without being a friend? OpenMS is a library and if that is our future proof data structure this should be possible for library users to implement this.

Reply:
Any tool/algorithm that adds data (which should be the majority) is fine.

Removing data is more difficult because of the consistency requirements. That is a side effect of having complex data connected by iterators. You cannot change this without losing the advantage of O(1) dereferencing. (Also, IdentificationData keeps look-up tables of object addresses for fast checking of reference validity, which naturally need to be kept up-to-date.)

Modifying stored data is unfortunately also difficult, because data in a `set` (or equivalently: multi-index container) is `const` (has to be to uphold the invariants of how elements are ordered). Multi-index offers mechanisms for updating elements (see https://www.boost.org/doc/libs/1_77_0/libs/multi_index/doc/tutorial/basics.html#ord_updating), the most efficient of which is `modify` - however, for that you need to implement a functor and ensure that the "key" of the element doesn't change. So this is not exposed as a general interface, but used to implement specific modification functions (e.g. `IdentificationData::addScore`).

The gist is this:
- We cannot allow users to remove elements because it could break the consistency guarantees.
- We cannot allow users to make general updates to elements because it could break container invariants.

With these constraints in mind, what would you suggest to improve the interface?
We could still add a more general updating mechanism using multi-index's `replace` (after checking that the "key" part of an element does not change), but I haven't found a compelling use case for this (that wasn't solved more efficiently using `modify`).

Reply:
I know it is a challenge. I am just looking at code I wrote and I often implement custom remove_if with lambda on PeptideHits and then call the usual removeUnreferenced updateGroups etc. to ensure that data is again consistent. This looks very C++ like. 
Brainstorming:
E.g could one one get additional unsafe access to delete and modify contents + mandatory cleanup/consistency check afterwards? Like e.g. unique_ptr that on destruction calls the consistency check / cleanup? Ideally wrappable in python

Reply:
We could add `filterXYZ(predicate, ...)` functions for each of the main data types (`XYZ`). They would use a user-supplied predicate (functor, lambda, ...) to determine whether to keep/remove each element, then call `cleanup` once at the end (with additional arguments to `cleanup` passed from `...`).

Reply:
The filter with predicate function is a step in the right direction. But we would need to consider:
- how to access the protected/private members?
- cleanup after every step seems a bit wasteful? Wouldn't that just allow AND combinations of filters between the different datatypes?
- how to e.g. filter PSMs based on properties in the different levels (e.g. remove all PSMs that match to a set of proteins with coverage > 90% OR are part of a protein group with some meta value )
- Shouldn't that be similar to a transaction? Do the checks only in the end?

Reply:
> how to access the protected/private members?

The filter functions would need to be in `IdentificationData` or a friend class (IDFilter).

> cleanup after every step seems a bit wasteful? Wouldn't that just allow AND combinations of filters between the different datatypes?

Much better than clean-up after removing a single element. Also, if you don't clean up between handling different data types, you may run into dangling references (e.g. if you remove proteins, then filter peptides by looking at their proteins).

> how to e.g. filter PSMs based on properties in the different levels (e.g. remove all PSMs that match to a set of proteins with coverage > 90% OR are part of a protein group with some meta value )

You always have read access to all the data, so complex filtering criteria aren't a problem.

In any case, can we get the merge done first, then build more filtering functionality?

Reply:
How would that access look like if we pass a lambda? E.g. how can you access private members in the lambda function that gets defined by a user and that gets passed to IDFilter?

Reply:
The predicate (e.g. lambda) would be applied to each data element (e.g. `IdentifiedPeptide`) and have access to its public members (which is usually all of them). The filter function would have privileged access to the container holding the elements, in order to remove those for which the predicate returns true (or false?).

Reply:
Ok so it Would be possible to query in a PSM filter lambda e.g. properties of the group the current considered PSM maps to? E.g. does the public interface of the ID datastructure makes this possible and not too inefficient?

Reply:
From a PSM you can follow a link to the peptide and from there to its protein(s). The protein groups you'd have to query separately using the public interface of IdentificationData (`getParentGroupSets`). Depending on your use case, you may want to populate a data structure that maps from protein to protein group for more efficient queries in that direction.

Reply:
and this would be possible by passing a lambda?

Reply:
Sure, just bind all the relevant variables that you need to access.

Reply:
ok but can I bind them? e.g. if they are private in the ID Datastructure.

Reply:
We're starting to turn in circles now. Of course you can't bind private members, but you don't need to. There are public, but immutable (`const`) accessors for all the data. (There's not much point storing data if you can't read it again.) The limitations only apply when you want to change stored data.

Reply:
Ok. Let's still wait for Julianus (back soon) as the affected LFQ workflow would have the biggest impact on his side